src/Integration/LocalServices/ErrorListInfoBarController.cs

```diff
@@ -21,22 +39,428 @@ public ErrorListInfoBarController(IHost host)
+        }
+
+        private enum BindingRequestResult { StartedUpdating, CommandIsBusy, RequestIsIrrelevant, NoActiveSection };
+
+        private class EventDrivenBindingUpdate
+        {
+            private readonly IHost host;
+            private readonly BoundSonarQubeProject binding;
+
+            public EventDrivenBindingUpdate(IHost host, BoundSonarQubeProject binding)
+            {
+                Debug.Assert(host != null);
+                Debug.Assert(binding != null);
+
+                this.host = host;
+                this.binding = binding;
+            }
+
+            private State.TransferableVisualState State
+            {
+                get { return this.host.VisualStateManager.ManagedState; }
+            }
+
+            private bool IsBusy
+            {
+                get
+                {
+                    return this.host.VisualStateManager.IsBusy;
+                }
+            }
+
+            public event EventHandler<BindingRequestResult> Finished;
+
+            public void ConnectAndBind()
+            {
+                if (this.host.ActiveSection == null)
+                {
+                    EventHandler activeSectionChanged = null;
+                    activeSectionChanged = (o, e) =>
+                    {
+                        this.host.ActiveSectionChanged -= activeSectionChanged;
+                        this.WhenNotBusy(this.Start);
+                    };
+                    this.host.ActiveSectionChanged += activeSectionChanged;
+                }
+                else
+                {
+                    this.WhenNotBusy(this.Start);
+                }
+
+                // Navigating at this point will work in both cases - if there's no active section
+                // we will navigate and the ActiveSectionChanged event will be triggered.
+                // If there's an active section will navigate (activate TE) and show the connecting/binding process.
+                // We drive the process via the ActiveSection, so this step is mandatory 
+                ITeamExplorerController teController = this.host.GetMefService<ITeamExplorerController>();
+                Debug.Assert(teController != null, "Cannot find ITeamExplorerController");
+                if (teController != null)
+                {
+                    teController.ShowSonarQubePage();
+                }
+
+            }
+
+            private void Start()
+            {
+                if (this.IsConnected(this.binding.ServerUri))
+                {
+                    // Skip the connection state
+                    this.UpdateBinding();
+                }
+                else
+                {
+                    // Start from connection state
+                    this.RefreshConnection();
+                }
+            }
 
+            private void RefreshConnection()
+            {
+                this.WhenNotBusy(this.ExecuteRefreshCommand);
```

Comment:
The longest nested chain is when already connected, in other cases since it will be busy it will break the nesting and start a new chain once no longer busy.